File: trapping-rain-water/trapping-rain-water.cpp
class Solution {
public:
  int trap(vector<int>& height) {
        int ans = 0, current = 0;
        stack<int> st;
        while(current < height.size()) {
            while(!st.empty() && height[current] > height[st.top()]) {
                int top = st.top();
                st.pop();
                if(st.empty()) break;
                int distance = current - st.top() - 1;
                int bh = min(height[current], height[st.top()]) - height[top];
                cout<<"height[current] "<<height[current]<<"   height[st.top()] "<<height[st.top()]<<endl;
                ans += distance * bh; 
            }
            st.push(current++);
        }
        return ans;
    }
};

File: trapping-rain-water/trapping-rain-water.js
/**
 * @param {number[]} height
 * @return {number}
 */
var trap = function(height) {
    const stack = [];
    let area = 0; 
    for(let i = 0; i < height.length; i++) {
        while(stack.length && height[stack[stack.length-1]] < height[i]) {
            const wallIdx = stack.pop();
            if(!stack.length) break;
            const currWidth = i - stack[stack.length-1] - 1;
            const currHeight = Math.min(height[stack[stack.length-1]], height[i]) - height[wallIdx];
            area += currWidth * currHeight;
        }
        stack.push(i);
    }
    return area;
};